### DISCUSSION TITLE
How to use a 3D TIFF image to calculate absolute permeability in OpenPNM?

### DISCUSSION THREAD

**Discussion by DOMFIRMUS** [PMEAL/OpenPNM#2947] [Q&A]
I am trying to use a 3D TIFF image as input to calculate absolute permeability in OpenPNM. I followed the example below, which uses a randomly generated cubic network to calculate permeability by applying the Stokes Flow algorithm. However, my goal is to replace this cubic network with the actual geometry of a rock sample obtained via tomography (3D TIFF image).

Example code used:

`import numpy as np
import openpnm as op
pn = op.network.Cubic(shape=[15, 15, 15], spacing=1e-6)
pn.add_model_collection(op.models.collections.geometry.spheres_and_cylinders)
pn.regenerate_models()
phase = op.phase.Phase(network=pn)
phase['pore.viscosity'] = 1.0
phase.add_model_collection(op.models.collections.physics.basic)
phase.regenerate_models()
inlet = pn.pores('left')
outlet = pn.pores('right')
flow = op.algorithms.StokesFlow(network=pn, phase=phase)
flow.set_value_BC(pores=inlet, values=1)
flow.set_value_BC(pores=outlet, values=0)
flow.run()
Q = flow.rate(pores=inlet, mode='group')[0]
A = op.topotools.get_domain_area(pn, inlets=inlet, outlets=outlet)
L = op.topotools.get_domain_length(pn, inlets=inlet, outlets=outlet)
K = Q * L / A
print(f'The value of K is: {K/0.98e-12*1000:.2f} mD')
`

![image](https://github.com/user-attachments/assets/0b51b801-b9ff-4cf7-bce2-259a3578985c)

![image](https://github.com/user-attachments/assets/bb77701f-fe4f-41f7-a14e-37ddba8bf9e1)

Question: What would be the correct procedure to import a 3D TIFF image and convert it into a network format usable in OpenPNM? I need help with the processing and modeling steps to adapt this image as input and calculate permeability as accurately as possible.

**Comment by jgostick** [comment]
You need to use a network extraction tool to get a network.  We have the snow algorithm in porespy.

**Reply by DOMFIRMUS** [reply]
ahh ok, thanks

**Reply by li-ruidong** [reply]
> You need to use a network extraction tool to get a network. We have the snow algorithm in porespy.

Hi, I have already got the network but still don't know how to calculate the tortuosity. I followed the example in https://openpnm.org/examples/applications/effective_diffusivity_and_tortuosity.html. However, there is no net.pores('left') or net.pores('right'). Do you know how to deal with this?

**Reply by Clifford-Wang** [reply]
> > 您需要使用网络提取工具来获取网络。我们在 porespy 中有 snow 算法。
> 
> 嗨，我已经有了网络，但仍然不知道如何计算迂曲度。我在 https://openpnm.org/examples/applications/effective_diffusivity_and_tortuosity.html 中效仿了这个例子。但是，没有 net.pores（'left'） 或 net.pores（'right'）。您知道如何处理这个问题吗？

It seems that in the latest algorithm, the keywords "net.pores('left')" or "net.pores('right')" have been modified to "xmin", "xmax", etc. You can print out "pn" to check what the keywords are, and then input appropriate parameters according to the printed information.
